backend/regalloc/regalloc_utils.ml

```diff
@@ -484,6 +330,59 @@ module StackSlots = struct
     done
 end
 
+module Substitution = struct
+  type t = Reg.t Reg.Tbl.t
+
+  let apply_reg : t -> Reg.t -> Reg.t =
+   fun subst old_reg ->
+    match Reg.Tbl.find_opt subst old_reg with
+    | None -> old_reg
+    | Some new_reg -> new_reg
+
+  let apply_array_in_place : t -> Reg.t array -> unit =
+   fun subst arr ->
+    for i = 0 to pred (Array.length arr) do
+      let old_reg = Array.unsafe_get arr i in
+      match Reg.Tbl.find_opt subst old_reg with
+      | None -> ()
+      | Some new_reg -> Array.unsafe_set arr i new_reg
+    done
+
+  let apply_array : t -> Reg.t array -> Reg.t array =
+   fun subst arr ->
+    let res = Array.copy arr in
+    apply_array_in_place subst res;
+    res
+
+  let apply_set : t -> Reg.Set.t -> Reg.Set.t =
+   fun subst set -> Reg.Set.map (fun reg -> apply_reg subst reg) set
+
+  let apply_instruction_in_place : t -> _ Cfg.instruction -> unit =
```

Comment:
In essence, it is the responsibility of the caller to call `Cfg_with_liveness.invalidate_liveness` when needed. I don't think it should be automatic because it is a bit like a transaction: you may apply substitutions in a bunch of places, also using the previous liveness, before triggering a recomputation.

Reply:
(Also, we are duplicating everything to not have to think about sharing.)

Reply:
> In essence, it is the responsibility of the caller to call `Cfg_with_liveness.invalidate_liveness` when needed. I don't think it should be automatic because it is a bit like a transaction: you may apply substitutions in a bunch of places, also using the previous liveness, before triggering a recomputation.

It makes sense. I think that functions that can invalidate liveness are currently not obvious from their description and should probably be marked so.

Reply:
Oh, in this case, it is the responsibility of the caller to apply consistent substituions to `arg` and `res`. My comment was about the sharing of the _arrays_ rather than the _registers_.